File: LC_JAVA/Sort_array_0_1_2.java
//{ Driver Code Starts
import java.io.*;
import java.util.*;

class Sort_array_0_1_2 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int t = Integer.parseInt(br.readLine());

        while (t-- > 0) {
            String input = br.readLine();
            String[] inputArray = input.split("\\s+");
            int a[] = new int[inputArray.length];

            for (int i = 0; i < a.length; i++) a[i] = Integer.parseInt(inputArray[i]);

            Solution ob = new Solution();
            ob.sort012(a);

            for (int num : a) {
                System.out.print(num + " ");
            }
            System.out.println();
        }
    }
}

// } Driver Code Ends


class Solution {
    // Function to sort an array of 0s, 1s, and 2s
    public void sort012(int[] arr) {
        // code here
        int p_0 = 0;
        int p_2 = arr.length-1;
        int p_1 = 0;
        int temp = 0;

        while(p_1 <= p_2){
            if(arr[p_1] == 1){
                p_1++;
            }
            else if(arr[p_1] == 0){
                temp = arr[p_1];
                arr[p_1] = arr[p_0];
                arr[p_0] = temp;
                p_0++;
                p_1++;
            }
            else if(arr[p_1] == 2){                
                temp = arr[p_1];
                arr[p_1] = arr[p_2];
                arr[p_2] = temp;
                p_2--;
            }
        }
    }
}

File: LC_PYTHON/Sort_array_0_1_2.py
class Solution:
    # Function to sort an array of 0s, 1s, and 2s
    def sort012(self, arr):
        # code here
        p0 = p1 = 0
        p2 = len(arr) - 1
        while p1 <= p2:
            if arr[p1] == 1:
                p1+=1
            elif arr[p1] == 0:
                arr[p1], arr[p0] = arr[p0], arr[p1]
                p0+=1
                p1+=1
            elif arr[p1] == 2:
                arr[p1], arr[p2] = arr[p2], arr[p1];
                p2-=1
         

#{ 
 # Driver Code Starts
def main():
    t = int(input().strip())  # Read the number of test cases
    ob = Solution()

    while t > 0:
        t -= 1
        arr = list(map(int,
                       input().strip().split())
                   )  # Read the array as space-separated integers
        ob.sort012(arr)  # Sort the array

        print(' '.join(map(str, arr)))  # Print the sorted array


if __name__ == "__main__":
    main()

# } Driver Code Ends